internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
+	// Review: not sure what all this pattern matching adds up to, the "singleton" pattern does an equality check
+	// ValidateCallerMatches(CallerPattern)
```

Comment:
`ValidateCallerMatches` is intended for methods that can be invoked more than one way, e.g., by some non-singleton actor A and also by Cron. The runtime will check that there's exactly one call to `ValidateCaller{Is|Matches}`, so we can't do this the natural way by just checking `Caller().Equals(A) || Caller().Equals(Cron)`.

Reply:
`ValidateCallerIs()` seems just like a simpler version of `ValidateCallerMatches()`, I understand the problem you are trying to solve. 

We can discuss the approach behind this methods in the previous comment.